### PR TITLE
Fix #5975

### LINKED ISSUE #5975: ./run ide watch displays white screen
Run the `./run ide watch` on a clean repository. A blank window appears.

This seems to be caused by hardcoding path to tailwind.css in https://github.com/enso-org/enso/commit/6f29262f90027995efbe8d5eee6d831496e97ad7#diff-57b07427fa0b3aa82b74d34fad7c62a14fb874dba1ca7a0e45cb86c67a021938R20

obraz

See https://github.com/enso-org/enso/pull/5972#issuecomment-1471916821

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
Fixes issue with blank screen caused by missing `tailwind.css`

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Review by PabloBuchu — APPROVED** [review]

File: app/ide-desktop/lib/content/package.json
Comment (wdanilo) [inline app/ide-desktop/lib/content/package.json:22]: Does it mean that tailwind CSS will be regenerated on every watch command? If so, this is bad, because we want watch to be as fast as possible, especially when chaining rust code.

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: app/ide-desktop/lib/content/package.json
Comment (somebody1234) [reply app/ide-desktop/lib/content/package.json:22]: yes, it is being regenerated on every watch command :(

Comment (somebody1234) [reply app/ide-desktop/lib/content/package.json:22]: actually, i have no clue. build times average 200ms for me when editing css/html

Comment (somebody1234) [reply app/ide-desktop/lib/content/package.json:22]: wait...

**Comment by somebody1234** [comment]
style plugin seems to currently be broken (not processing any files)

**Comment by PabloBuchu** [comment]
@somebody1234 generating tailwind.css can't add time to watch iteration. We also can't use plugins (because plugins runs every time on watch). 

We need to add a logic to `esbuild-config.ts` that calls `cd ../dashboard/src/authentication && npx tailwindcss -i src/styles/index.css -o ../../../../../../dist/gui/assets/tailwind.css -c tailwind.config.ts --minify`. But the function need to be called once

File: app/ide-desktop/lib/content/package.json
Comment (somebody1234) [reply app/ide-desktop/lib/content/package.json:22]: should be fixed - i've added an inline plugin that only registers `onStart`, which iiuc should only run once

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (mwu-tow) [inline app/ide-desktop/lib/content/esbuild-config.ts]: Please avoid using the `exec*('command')` form whenever possible. Always prefer `exec*('program', [args])`. Otherwise, it is nearly impossible to support things like spaces in paths in a portable manner.

Also, please extract long, repetitive expressions like `path.resolve(THIS_PATH, 'src', 'tailwind.css')` to named variables. That will improve code clarity.

**Review by mwu-tow — CHANGES_REQUESTED** [review]

**Comment by mwu-tow** [comment]
@PabloBuchu 
> We also can't use plugins (because plugins runs every time on watch).

Are you sure? I have no idea myself, frankly speaking, but I thought that if a plugin is registered only for cases when esbuild loads or resolves file. Which I hope does not happen for all files every time the watch catches something? 
I mean… that'd be so wrong, especially for a tool that boasts of its great performance.

**Comment by PabloBuchu** [comment]
> @PabloBuchu
> 
> > We also can't use plugins (because plugins runs every time on watch).
> 
> Are you sure? I have no idea myself, frankly speaking, but I thought that if a plugin is registered only for cases when esbuild loads or resolves file. Which I hope does not happen for all files every time the watch catches something? I mean… that'd be so wrong, especially for a tool that boasts of its great performance.

@mwu-tow tbh I wasn't sure on that but I got that from talking with @wdanilo I was also surprised.. we need to check then

**Comment by somebody1234** [comment]
so it seems like `onResolve` is triggered by import resolution, and `onLoad` is triggered by loading an import (which makes sense because esbuild is a js build tool, not a general build tool). if we don't import the css, they will never fire

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (somebody1234) [reply app/ide-desktop/lib/content/esbuild-config.ts]: changed to `spawnSync(program, args)`; extracted to named variables

File: app/ide-desktop/lib/content/package.json
Comment (wdanilo) [reply app/ide-desktop/lib/content/package.json:22]: > actually, i have no clue. build times average 200ms for me when editing css/html

This would not be a problem if this was a simple build script. But 200ms here and there can cause big problems for our dev speed. So we take a lot of care about it. Thanks for fixing it! Running it once seems like perfect solution.

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (wdanilo) [reply app/ide-desktop/lib/content/esbuild-config.ts]: > changed to `spawnSync(program, args)`; extracted to named variables

Please double check that this process is killed when parent process is killed. We had issues with spawn in the past. CC @mwu-tow

**Comment by wdanilo** [comment]
Ok, when this will be fully checked and resolved, ping me pls and Id love to do fast review then.

**Comment by somebody1234** [comment]
so i was looking into re-running postcss only when the css file changed, but it seems like esbuild's incremental mode doesn't really work with plugins - so i think we'll have to stick with this solution for the near future. so i think i'll just fix the merge conflict

**Comment by somebody1234** [comment]
not sure why github UI said the merge was too complex to do through the web UI, because it merged cleanly locally

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (somebody1234) [reply app/ide-desktop/lib/content/esbuild-config.ts]: unfortunately this does seem to be a problem - i've tested by switching from `spawnSync` to `spawn`, adding a `process.exit(1)`, and checking if the child process is still running. i *think* avoiding `npx` fixes it, but the current solution i've ended up with is quite hacky

**Comment by somebody1234** [comment]
@wdanilo i think all issues have been resolved now

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/esbuild-config.ts]: Please refactor these to variable / variables to make the expressions one-liners instead (see styling code guide for more info).

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/esbuild-config.ts]: duplicated code line. Exactly the same logic was added above. Please refactor it to a common constant somewhere close to the top of the file.

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/esbuild-config.ts]: 1. Sentences in logs should be real English sentences. A dot is missing here.
2. In such logs we normally enclose paths in quotes (`'path'`), so in case something broke and e.g. source is an empty string its easier to debug them.

File: app/ide-desktop/lib/content/tailwind.config.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/tailwind.config.ts:5]: That's a very strange dependency here. We have `content` which does not use tailwind (is that correct?) and it refers another library `dashboard` which uses tailwind to build the styles? If Im correct, then the building of tailwind should be integrated in the build script of `dashboard` maybe? Or maybe this should be something more top-level? I don't understand why `content` is now responsible of building artifacts for other libraries. CC @PabloBuchu  and @mwu-tow - how should it be done in the correct way, do you have some thoughts here?

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (somebody1234) [reply app/ide-desktop/lib/content/esbuild-config.ts]: what exactly should i refactor to variables?

File: app/ide-desktop/lib/content/tailwind.config.ts
Comment (somebody1234) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: i believe currently `content` contains the desktop IDE frontend, whereas `dashboard` contains the dashboard, which will be used for both cloud and desktop.

(so, if the dashboard is supposed to be in the desktop IDE, then it should be here too)

on second thought, this might be a bad solution. because the css is no longer in the frontend, and that might be an issue...

Comment (somebody1234) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: maybe now is a good time to figure out how to *properly* structure `content` and `dashboard`

Comment (wdanilo) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: > i believe currently `content` contains the desktop IDE frontend, whereas `dashboard` contains the dashboard, which will be used for both cloud and desktop.

It is correct but this is VERY temporary solution. Content is meant to be run both in Electron and in the cloud. It is not being used in cloud only because there was no time to refactor it, but it should change ASAP, as now it is in a very confusing state. Dashboard is thus dependency of `content`, as it is displayed as part of `content`. `Dashboard` is a separate library, in a separate folder, next to `content`, and `content` imports it, which is IMO correct.

Comment (somebody1234) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: about merging the code - maybe we could consider using vite to replace esbuild + create-react-app
vite:
- *does not* support esbuild plugins because the architecture is incompatible with their own plugins - so we'll have to look for alternatives. this shouldn't be too hard because vite has a pretty large ecosystem
- supports HMR (we are using this to speed up testing the cloud frontend)
- supports tailwind
- seems to support electron

importantly that means we could build the dashboard using the same build tool as content, which should make it a lot easier for content to depend on dashboard

Comment (wdanilo) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: We've chosen `esbuild` because of one reason - fast build times. It is crucial for us to maintain as small build time as possible after changing Rust sources. Do you know how does vite compare to it?

Comment (somebody1234) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: vite uses esbuild for transpilation, but it uses rollup fpr bundling which is written in js so it's not the fastest.

as an alternative to vite we could look into manually implementing some HMR support in development mode, or just not using HMR completely? because esbuild *does* support live reloads out of the box, it's just HMR that it doesnt

the downside of live reloading as compared to HMR is that the cloud frontend would send requests to the backend every time the page is live-reloaded, but that may not be much of a problem for us

Comment (somebody1234) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: note that not using HMR should barely affect desktop IDE, since it is not sending requests to the backend (no network requests = latency for e.g. listing projects is a lot lower)

Comment (wdanilo) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: Could you chat about this with @mwu-tow, please? He was setting up `esbuild` here and he has the best knowledge of how the build system works / should work. Probably catching him on Discord or making a fast call would be the fastest way forward.

Comment (PabloBuchu) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: @wdanilo This is a problem with single `index.html` and tailwind.css dependency. Styles are being used in `/dashboard` app but `tailwind` itself (and config) is being built in content since it is built by esbuilt. 

In dashboard we don't have any place / building scripts to do that since it is the content that also runs dashboard/authentication

Happy to jump in on call and quick chat on this since this PR fixes for us an important bug

Comment (mwu-tow) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: So to answer — I am not knowledgeable about web development nor the whole js, ts, react ecosystems. `esbuild` was chosen because it was much faster than `webpack` (which we used previously) and covered our needs. IIRC, at the time of that decision, we hadn't expected that any subproject that requires anything not supported by `esbuild` would be added. The dashboard was planned to be written in Rust.

Are there other problems with using `esbuild` for `dashboard` apart from HMR? Why was `create-react-app` that uses `webpack` under the hood introduced after we decided to replace all `webpack` with `esbuild`? Was there an explicit decision for this choice, or was it just a default that persisted?

My quick thoughts:
* If both `content` and `dashboard` require `tailwind.css`, then both of them should have it as their dependency, and their build systems should support it in their respective ways.
* We should not care about deduplicating this tailwind generation by ourselves; this should be something that build system (being aware of the whole dependency tree) should handle. In the worst-case scenario, setting the file twice is not a disaster.
* `content` should not assume that `dashboard` uses tailwind; it is an implementation detail.
* It would be good to switch to `esbuild` for `dashboard` without too much effort.
* If something like `vite` can replace `esbuild` for all our projects and save us some manual tinkering over `esbuild`, then it is even better, but I'm concerned about the necessary work to migrate to it.

Comment (somebody1234) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: > are there any other problems?

i don't think so. if there are any, i think they will be minor enough that they can and should be solved now rather than later

> why was `create-react-app` used even though it uses `webpack`?

i assume it's just most people's go-to template - so yes, a default that persisted

> If both `content` and `dashboard` require `tailwind.css`

well, `content` requires/will require `dashboard` which in turn requires `tailwind.css`

i think i'll try switching to esbuild for `dashboard`. if that works i'll probably open up a separate PR

Comment (PabloBuchu) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: @wdanilo @mwu-tow can we transform this path issue into a FIXME? We will refactor how the dashboard and content is handled after we clean up the rest of the merges.

Comment (indiv0) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: > Why was `create-react-app` that uses `webpack` under the hood introduced after we decided to replace all `webpack` with `esbuild`?

Yes, this was a default that persisted. Before `dashboard` was added as a `content` dependency, it was using `create-react-app`. There is no reason not to switch, aside from migration efforts.

> I don't understand why content is now responsible of building artifacts for other libraries.

`content` should be building `tailwind.css`. `content` will be our sole entrypoint. `dashboard` also builds `tailwind.css` only so we can run the `dashboard` separately. This gives us HMR and faster dev times.

**TL;DR:** The solution in this PR makes sense to me, with our "one entrypoint" approach. Switching to `esbuild` is a good idea but shouldn't block this PR from going in, since it fixes our dashboard UI and lets us start testing Electron+Dashboard+IDE together.

Comment (wdanilo) [reply app/ide-desktop/lib/content/tailwind.config.ts:5]: I agree 100% with what @mwu-tow has written. Btw, Michal, thanks for breaking that down so clearly, as always ❤️ 

Anyway, if content do not uses tailwind, and probably will never use it, the tailwind building should not be integrated with content, it should be integrated with `dashboard`. I think the idea to move dashboard to `esbuild` and keep every part of our TS/JS ecosystem using the same build tools is the right choice @indiv0. Then, we can just integrate building tailwind in the `dashboard` `esbuild` config, and use `dashboard` as a library in `content`, while keeping `content` unaware of internal implementation details of `dashboard`. 

@PabloBuchu I'm OK with merging this as is iff we will be fixing that in the following PRs. This introduces pretty bad dependency between packages and makes the tech debt much worse. So yes, we can move forward, but let's please fix it as @mwu-tow, @indiv0, and @somebody1234 proposed above :)

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/esbuild-config.ts]: I think my comment from my previous PR review was lost again here. I added here a comment that this is a really long expression and according to our style guide, we should split it into one-line exprs instead. I'd be thankful if we will not use anymore force pushes. TBH, I do not understand why they are used - for the last 10? years, I never used a force push in my PRs, so I kind of don't understand what forces us to use them here :)

**Review by wdanilo — APPROVED** [review]
Accepting to move forward, however:
1. Please apply the code styling comment.
2. Let's fix the dependency issue in the next PR :)

File: app/ide-desktop/lib/content/esbuild-config.ts
Comment (somebody1234) [reply app/ide-desktop/lib/content/esbuild-config.ts]: nope, it's [still here](https://github.com/enso-org/enso/pull/5976#discussion_r1142852535). there have been no force pushes since that comment.

i've been force pushing when:
- the commit history has diverged too much (between the old base branch, and the new base branch), so redoing the changes is much easier, or
- it's a minor formatting change, and it's early enough (say, 5 minutes) that there are no reviews, or
- i did it through a github web UI with a rebase. this is probably a mistake, i'll be using regular merges through the web UI instead when i do thi

Comment (somebody1234) [reply app/ide-desktop/lib/content/esbuild-config.ts]: but yeah i'm still not sure how exactly to split this into one-line exprs - do i split the array into two parts? not sure that will be more readable...

Comment (wdanilo) [reply app/ide-desktop/lib/content/esbuild-config.ts]: Sure! What about this?

```
const tailwind_opts = [TAILWIND_BINARY_PATH, '-i', TAILWIND_CSS_PATH]
const output_opts = ['-o', dest]
const config_opts = ['-c', config, '--minify']
childProcess.spawnSync(`node`, [tailwind_opts, output_opts, config_opts].flat())
```

Comment (somebody1234) [reply app/ide-desktop/lib/content/esbuild-config.ts]: still too long, unfortunately... (the last line is 102 characters.)

i guess i'll just try extracting it to a top level variable

Comment (wdanilo) [reply app/ide-desktop/lib/content/esbuild-config.ts]: Don't spend too much time on it. If it would not fit, lets just skip it :)

**Comment by somebody1234** [comment]
i've tested a bit - turns out it wasn't properly working before (it ran on every rebuild as well), it should be fixed now.
reviews should all be addressed, but it would be good to double check.

**Review by mwu-tow — APPROVED** [review]

**Pull request merged**
